apps/admin/backend/src/store.ts

```diff
@@ -353,45 +360,80 @@ export class Store {
   /**
    * Adds a CVR file entry record and returns its ID. If a CVR file entry with
    * the same contents has already been added, returns the ID of that record and
-   * associates `cvrFileId` with it.
+   * merely associates `cvrFileId` with it.
    */
-  addCastVoteRecordFileEntry(
-    electionId: Id,
-    cvrFileId: Id,
-    ballotId: BallotId,
-    data: string
-  ): Result<
+  addCastVoteRecordFileEntry({
+    electionId,
+    cvrFileId,
+    ballotId,
+    metadata,
+    votes,
+  }: {
+    electionId: Id;
+    cvrFileId: Id;
+    ballotId: BallotId;
+    metadata: CastVoteRecordMetadata;
+    votes: string;
+  }): Result<
     { cvrId: Id; isNew: boolean },
     {
-      kind: 'BallotIdAlreadyExistsWithDifferentData';
-      newData: string;
-      existingData: string;
+      type: 'ballot-id-already-exists-with-different-data';
     }
   > {
     const existingCvr = this.client.one(
       `
         select
           id,
-          data
+          ballot_style_id as ballotStyleId,
+          ballot_type as ballotType,
+          batch_id as batchId,
+          batch_label as batchLabel,
+          precinct_id as precinctId,
+          scanner_id as scannerId,
+          sheet_number as sheetNumber,
+          votes as votes
         from cvrs
         where
           election_id = ? and
           ballot_id = ?
       `,
       electionId,
       ballotId
-    ) as { id: Id; data: string } | undefined;
+    ) as
+      | {
+          id: Id;
+          ballotStyleId: string;
+          ballotType: string;
+          batchId: string;
+          batchLabel: string;
+          precinctId: string;
+          scannerId: string;
+          sheetNumber: number | null;
+          votes: string;
+        }
+      | undefined;
```

Comment:
Here we're just retrieving all the data about a record in order to compare it to the new record. I wonder if it would be better to just hash all records and compare to a hash? Slightly higher space requirements, maybe slightly lower time requirements (this dupe checking doesn't take very long in perf testing), but this code becomes more readable. Based on those tradeoffs, decided not to pursue the hashing now, but maybe there's a case for it.